lib/srv/discovery/discovery.go

```diff
@@ -1327,8 +1322,81 @@ func (s *Server) getAllGCPServerFetchers() []server.Fetcher {
 	return allFetchers
 }
 
+// acquireDiscoveryGroup tries to acquire a lock if the Service has a DiscoveryGroup
+// It will retry using an exponential backoff algorithm.
+func (s *Server) acquireDiscoveryGroup() error {
+	if s.DiscoveryGroup == "" {
+		return nil
+	}
+
+	retry, err := retryutils.NewRetryV2(retryutils.RetryV2Config{
+		First:  0,
+		Driver: retryutils.NewExponentialDriver(defaults.HighResPollingPeriod),
+		Max:    defaults.LowResPollingPeriod,
+		Jitter: retryutils.NewHalfJitter(),
+		Clock:  s.clock,
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	var lease *services.SemaphoreLock
+
+	s.releaseGroupLockFn = func() {
+		if lease != nil {
```

Comment:
From what I can tell there's a potential race condition on `s.releaseGroupLockFn` and a potential race condition on `lease`.

The logical race condition of `(*Server).Stop` getting called right before or during a call to `acquireDiscoveryGroup`, leading to resources that are never going to get cleaned up is also still there I think?

Reply:
Can you please take another look?
I think I fixed both races, and `go test -race ...` seems to agree.